### PR TITLE
Liel/new message notifications

### PR DESCRIPTION
**Pull request by lielsa2023**

### CODE REVIEW

**Review by ayeletBinder — APPROVED** [review]

File: client-side/src/stories/header/Icon.jsx [hunk 1]
```diff
@@ -1,27 +1,33 @@
...
 import { MailOutline, Mail } from '@mui/icons-material';
+import useWebSocket from '../../webSocket';
+import messages from '../../components/messages/messages'
 import './icon.scss';
 
-const MessageIcon = ({ messages = [] }) => {
-  let cntUnreadMessages = messages.filter(message => !message.read).length;
+const MessageIcon = () => {
+  const [openMesagges,setOpenMesagges] = useState(false);
+  const userId = '66961bef8ca0916dcbfdabc3'; //change the userId according the correct user...
```
Comment (SariGlick) [inline client-side/src/stories/header/Icon.jsx:10]: remove static value, get current user from redux store

File: client-side/src/stories/header/Icon.jsx [hunk 2]
```diff
@@ -1,27 +1,33 @@
...
+
   const hasUnreadMessages = cntUnreadMessages > 0;
 
   const handleClick = () => {
     // Displays a generic component with the list
   };
 
   return (
-    <IconButton onClick={handleClick} className="iconButton">
+    <IconButton onClick={()=>{setOpenMesagges(!openMesagges);}} className="iconButton">
```
Comment (SariGlick) [inline client-side/src/stories/header/Icon.jsx:20]: remove ;

File: client-side/src/webSocket.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  const [socket, setSocket] = useState(null);
+  const [initialLoad, setInitialLoad] = useState(true);
+  const { enqueueSnackbar } = useSnackbar(); 
+  
+
+  useEffect(() => {
+    const ws = new WebSocket('ws://localhost:8080');
+
+    ws.onopen = () => {
+      console.log('Connected to WebSocket server');
```
Comment (SariGlick) [inline client-side/src/webSocket.js:17]: remove logs

File: client-side/src/webSocket.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      const response =event.data;            
+      if (!initialLoad && response > cntUnreadMessages) {
+        enqueueSnackbar(<ToastMessage open={true} message={'You received a new message'} type="info" />); 
+      }      
+      setCntUnreadMessages(response);
+      setInitialLoad(false);
+    };
+
+    ws.onclose = () => {
+      console.log('Disconnected from WebSocket server');
```
Comment (SariGlick) [inline client-side/src/webSocket.js:31]: remove log

File: server-side/app.js [hunk 5]
```diff
@@ -4,34 +4,35 @@ import cors from 'cors';
...
-app.use('/users', usersRouter);
-app.use('/api', settingsRouter);
-
+app.use('/uploads',express.static('uploads'))
+app.use('/preferences',preferencesRouter);
+app.use('/websites',websitesRouter);
+app.use('/profiles',profilesRouter);
+app.use('/vistedWebsites',visitedWebSitesRouter);
+app.use('/users',usersRouter);
+app.use('/message',MessageRouter);
```
Comment (SariGlick) [inline server-side/app.js:34]: notice that you delete /api route, please return it

File: server-side/controllers/message.controller.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
+
+
+import mongoose from 'mongoose';
+import Message from '../models/Message.model.js';
+import MessageType from '../models/MessageType.model.js';
+import { eventEmitter } from '../webSocket.js'; // Import event emitter
+
+export const getMessages = async (req, res) => {
+  try {
+    debugger
```
Comment (SariGlick) [inline server-side/controllers/message.controller.js:10]: remove debugger

File: server-side/controllers/message.controller.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+     const messages = await Message.find().select('-__v');
```
Comment (SariGlick) [inline server-side/controllers/message.controller.js:11]: move all DB calls to separated service file

File: server-side/controllers/message.controller.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    res.status(200).json(messages);
+  } catch (error) {
+    res.status(500).json({ error: error.message });
+  }
+};
+
+export const getMessageById = async (req, res) => {
+  try {
+
+    const message = await Message.findById(req.params.id,{__v:0}) //.populate('type').populate('userId');
```
Comment (SariGlick) [inline server-side/controllers/message.controller.js:22]: remove comment

File: server-side/router/message.router.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import  express from 'express';
+
+import { getMessages,getMessageById,addMessage, updateMessage, deleteMessage,getMessagesByUserId,  } from '../controllers/message.controller.js';
+const MessageRouter=express.Router();
+
+MessageRouter.get('/',getMessages);
+MessageRouter.get('/:id', getMessageById);
+// MessageRouter.post('/count/:id', getCountUnreadMessages);//???
```
Comment (SariGlick) [inline server-side/router/message.router.js:8]: remove comment

File: server-side/webSocket.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import { WebSocketServer } from 'ws';
+import mongoose from 'mongoose';
+import { connectMongo } from './config/db.js'; // Assuming this connects to MongoDB
+import { EventEmitter } from 'events';
```
Comment (SariGlick) [inline server-side/webSocket.js:4]: reorder the imports

File: server-side/webSocket.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+dotenv.config();
+
+const wss = new WebSocketServer({ port: 8080 });
+
+connectMongo();
+
+const eventEmitter = new EventEmitter(); 
+
+wss.on('connection', (ws) => {
+  console.log('Client connected');
```
Comment (SariGlick) [inline server-side/webSocket.js:16]: remove all logs

File: server-side/webSocket.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  ws.on('message', async (message) => {
+    ws.userId = message.toString();
```
Comment (SariGlick) [inline server-side/webSocket.js:18]: you should add switch-case to message.type,  ws can treat multiple types of messages

**Review by SariGlick — CHANGES_REQUESTED** [review]

**Review by SariGlick — APPROVED** [review]

**Pull request merged**